### ISSUE TITLE
Jersey core client 3.1.9 memory leak

### ISSUE THREAD

**Issue by jonkjenn** [eclipse-ee4j/jersey#5797]
If I'm not missing something this is a memory leak if you target unique URLs.

https://github.com/eclipse-ee4j/jersey/blob/f35691f4e4f1e6440ba2d384de798c4efdb2f932/core-client/src/main/java/org/glassfish/jersey/client/HttpUrlConnectorProvider.java#L301

**Comment by MarioSob** [comment]
This is indeed a memory leak. This map can only increase in size.

I suggest this class to take the following shape:
```
private static class DefaultConnectionFactory implements ConnectionFactory {
    private final ConcurrentHashMap<URL, Semaphore> locks = new ConcurrentHashMap<>();

    @Override
    public HttpURLConnection getConnection(URL url) throws IOException {
        return connect(url, null);
    }

    @Override
    public HttpURLConnection getConnection(URL url, Proxy proxy) throws IOException {
        return connect(url, proxy);
    }

    private HttpURLConnection connect(URL url, Proxy proxy) throws IOException {
        Semaphore thisThreadLock = new Semaphore(0);
        Semaphore previousThreadLock = locks.putIfAbsent(url, thisThreadLock);
        try {
            if (previousThreadLock != null) {
                previousThreadLock.acquireUninterruptibly();
            }
            return proxy == null
                    ? (HttpURLConnection) url.openConnection()
                    : (HttpURLConnection) url.openConnection(proxy);
        } finally {
            locks.remove(url, thisThreadLock);
            thisThreadLock.release();
        }
    }
}
```

**Comment by jansupol** [comment]
Thank you for reporting this. We have a fix in progress: #5794
